src/fparser/Fortran2003.py

```diff
@@ -561,32 +561,51 @@ class BracketBase(Base):
 ::
     <bracket-base> = <left-bracket-base> <something> <right-bracket>
     """
+    @staticmethod
     def match(brackets, cls, string, require_cls=True):
+        ''' The generic match method for all types of bracketed
+        expressions '''
         bracket_len = len(brackets)/2
         left = brackets[:bracket_len]
         right = brackets[-bracket_len:]
-        if string.startswith(left) and string.endswith(right):
-            # We may have something like "(a + b)*(a - b)" so have
-            # to check - we start with one open bracket. If we reach
-            # zero open brackets before we get to the end then the
-            # opening bracket at the start of the string does not
-            # correspond to the closing bracket at the end of it.
-            num_open = 1
-            for idx in range(bracket_len, len(string)-bracket_len):
-                if string[idx:idx+bracket_len] == left:
-                    num_open += 1
-                elif string[idx:idx+bracket_len] == right:
-                    num_open -= 1
-                if num_open == 0:
-                    return
-            line = string[bracket_len:-bracket_len].strip()
-            if not line:
-                if require_cls:
-                    return
-                return left, None, right
-            return left, cls(line), right
-        return
-    match = staticmethod(match)
+
+        if not (string.startswith(left) and string.endswith(right)):
+            return
+
+        # Check whether or not there's anything between the open
+        # and close brackets
+        line = string[bracket_len:-bracket_len].strip()
+        if not line:
+            if require_cls:
+                return
+            return left, None, right
+
+        # There's some content between the open and close brackets.
+        # We may have something like "(a + b)*(a - b)" so have
+        # to check - we start with one open bracket. If we reach
+        # zero open brackets before we get to the end then the
+        # opening bracket at the start of the string does not
+        # correspond to the closing bracket at the end of it.
+        # Unless of course any interim brackets we encounter are
+        # within strings...
+        num_open = 1
+        in_string = False
+        for idx in range(bracket_len, len(string)-bracket_len):
+            if string[idx] == '"' or string[idx] == "'":
```

Comment:
There is a potential problem here. Quotes may be escaped in strings which would cause this logic to break. For example "\"". I think we need more general support for recognising strings in the parser and this is a separate issue. I have therefore created a separate issue for this #18. Note, this code is therefore known to contain a bug but is a significant improvement on before so the bug is being punted to another ticket.